Zparks/8 9 24/update csg for multi project

ZachCMP opened the pull request:
[//]: # 'remove this if PR is for a release-* branch'
# _Please squash merge this PR_

## Description

Updates CSG import to allow importing multiple projects to a single CSG program. Also adds a test for the reporting logic

## Type of change
[//]: # 'remove options that are not relevant'
- [x] New feature (adds functionality)

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

ZachCMP committed:
Fixup CSG to handle multiple projects to a single program

ZachCMP committed:
Add reporting test

ZachCMP committed:
Merge remote-tracking branch 'origin/release-129' into zparks/8-9-24/update-csg-for-multi-project

ZachCMP commented:
@eanders Looks like some other changes came through for the `warehouse.sql` file. Let me know if you want me to clean them up, or if this looks okay to you

eanders commented on drivers/ma_reports/app/models/ma_reports/csg_engage/report_components/program.rb:
```diff
@@ -26,12 +26,12 @@ def initialize(program_mapping)
 
     def households_scope
-      project.enrollments.heads_of_households.preload(project: [:project_cocs])
+      GrdaWarehouse::Hud::Enrollment.joins(:project).where(project: { id: project_ids }).heads_of_households.preload(project: [:project_cocs])
```
Where is household_scope getting used?  This change potentially will give you back households with the same `HouseholdID` in different projects in the same scope, so that should be disambiguated with `[HouseholdID, ProjectID, data_source_id]` if you are counting households.  That said, HouseholdID is _supposed_ to be unique across a data source, so that's a theoretical need.

ZachCMP committed:
Adjust HH identifier

ZachCMP committed:
Switch to using hex for ids

ZachCMP commented:
@eanders This should be good to go now. I adjusted the household identifier to be an aggregate key of the enrollment id, project id, and data source id. To keep the IDs somewhat anonymized I converted them to hex before constructing the aggregate key. Let me know if you need anything else from me to get this on staging

ZachCMP committed:
Merge remote-tracking branch 'origin/release-129' into zparks/8-9-24/update-csg-for-multi-project

eanders commented on drivers/ma_reports/app/models/ma_reports/csg_engage/report_components/household.rb:
```diff
@@ -12,7 +12,9 @@ def initialize(hoh_enrollment)
 
-    field('Household Identifier') { hoh_enrollment.id.to_s }
+    field('Household Identifier') do
+      [hoh_enrollment.id, hoh_enrollment.project.id, hoh_enrollment.data_source.id].map { |id| id.to_s(16) }.join('::')
```
So, in this case using `.id` on the enrollment would be sufficient, but I _think_ you want to tie it to eh `HouseholdID` field in case they change who is HoH.  Using `project.id` also means you don't need `data_source.id`.

We usually do something like this:
```suggestion
      Digest::MD5.hexdigest([hoh_enrollment.HouseholdID, hoh_enrollment.project.id].join('|'))
```

That should give a consistent value based on the project and HouseholdID and should be the correct length to fit in the HouseholdID field.

ZachCMP replied:
@eanders The field length limit is 20 chars, and I think MD5 hashes are 32. I guess we could just truncate it, but technically that could cause collisions, though unlikely.

ZachCMP replied:
@eanders This is fixed now

eanders commented on drivers/ma_reports/spec/models/ma_reports/csg_engage/report_components/report.rb:
Should this file end in `_spec.rb`?  I'm not sure our CI will catch this without that naming convention.

ZachCMP replied:
Fixed

eanders reviewed (changes requested):
Just a couple comments/questions.  Otherwise, I think this should be good.

ZachCMP committed:
Update file names

ZachCMP committed:
Fixup household ID logic
